### PR TITLE
node: Update Node Monitor Grace Period default duration to 50s

### LINKED ISSUE #121793: HTTP2_PING_TIMEOUT_SECONDS+HTTP2_READ_IDLE_TIMEOUT_SECONDS > 40s will cause all node NotReady
### What happened?

The controller-manager node-lifecycle-controller watches node lease and update `nodeHealthMap` probeTimestamp , and it check node probeTimestamp is before `now - node-monitor-grace-period`. The node-monitor-grace-period default value is 40s.

When the client connection lost, the lease watcher cannot receive new event. With the http2 health check,   HTTP2_PING_TIMEOUT_SECONDS default is 30s, HTTP2_READ_IDLE_TIMEOUT_SECONDS default is  15s, it will delay 45s to kill http2  connection.  

During this period, node-lifecycle-controller will mark all nodes NotReady.

Here are some logs

```
I1103 23:00:28.464015       1 node_lifecycle_controller.go:1097] node 10.0.10.240 hasn't been updated for 40.010894415s. Last Ready is: &NodeCondition{Type:Ready,Status:True,LastHeartbeatTime:2023-11-03 22:57:18 +0800 CST,LastTransitionTime:2023-04-21 14:39:17 +0800 CST,Reason:KubeletReady,Message:kubelet is posting ready status,}

W1103 23:00:41.408493       1 reflector.go:441] k8s.io/client-go/informers/factory.go:134: watch of *v1.Lease ended with: an error on the server ("unable to decode an event from the watch stream: http2: client connection lost") has prevented the request from succeeding
```

### What did you expect to happen?

When the http2 connection lost, the watch should stop as soon as possible. The reflector should relist resource, and node-lifecycle-controller should not mark all node NotReady.

HTTP2_PING_TIMEOUT_SECONDS+HTTP2_READ_IDLE_TIMEOUT_SECONDS should be less than 40s.

### How can we reproduce it (as minimally and precisely as possible)?

1. Find the controller-manager node-lifecycle-controller watch client srcIP:srcPort. Drop all tcp package send into srcIP:srcPort.
2. Wait for more than 40 seconds, all node will be marked NotReady
3. All node will be Ready after some seconds.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# Server Version: version.Info{Major:"1", Minor:"28+", GitVersion:"v1.28.3", GitCommit:"", GitTreeState:"clean", BuildDate:"2023-10-19T07:10:40Z", GoVersion:"go1.20.10", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by devppratik**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Ref: [121793](https://github.com/kubernetes/kubernetes/issues/121793)

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/121793 

#### Does this PR introduce a user-facing change?

```release-note
The default value for node-monitor-grace-period has been increased to 50s (earlier 40s) (Ref - https://github.com/kubernetes/kubernetes/issues/121793)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by rtheis — COMMENTED** [review]
Please update https://github.com/kubernetes/kubernetes/blob/2171bcb78966539a0c66b5ae8d9a96f9277a7e82/pkg/controller/nodelifecycle/node_lifecycle_controller_test.go#L55 too.

**Comment by aojea** [comment]
/lgtm

/assign @dchen1107 
/hold

@liggitt do we have any criteria or precedence to change defaults on components? In this case, if you read the associated issue the existing default is clearly wrong as it ends with an unreliable configuration for the cluster

File: pkg/controller/nodelifecycle/config/v1alpha1/defaults.go
Comment (liggitt) [inline pkg/controller/nodelifecycle/config/v1alpha1/defaults.go:41]: The linked issue does a great job explaining *why* the current default is not safe, but the docs here or on the flag don't help users at all. This is now a parameter which can be tuned which has some settings we *know* are dangerous that we're not detecting / warning about at all.

At the very least, we need comments here and in the flag help explaining that this should be greater than HTTP2_PING_TIMEOUT_SECONDS (default 30) + HTTP2_READ_IDLE_TIMEOUT_SECONDS (default 15), and why.

Those values are *also* tunable... at startup, should we check that the effective values for the HTTP2 timeouts and this grace period are coherent and warn if they are not?

**Comment by liggitt** [comment]
> @liggitt do we have any criteria or precedence to change defaults on components? In this case, if you read the associated issue the existing default is clearly wrong as it ends with an unreliable configuration for the cluster

I think this surfaces as a flag for the controller manager, there's no config file yet. We've changed defaults in the past when they were problematic.

File: pkg/controller/nodelifecycle/config/v1alpha1/defaults.go
Comment (dchen1107) [reply pkg/controller/nodelifecycle/config/v1alpha1/defaults.go:41]: +1

**Comment by dchen1107** [comment]
Thanks for addressing the issue. LGTM. Please address @liggitt's comment above before approving it. cc/ @wangzhen127 for vis.

File: pkg/controller/nodelifecycle/config/v1alpha1/defaults.go
Comment (devppratik) [reply pkg/controller/nodelifecycle/config/v1alpha1/defaults.go:46]: > The linked issue does a great job explaining why the current default is not safe, but the docs here or on the flag don't help users at all. This is now a parameter which can be tuned which has some settings we know are dangerous that we're not detecting / warning about at all.

I have added a few comments and updated the flag description as well

> Those values are also tunable... at startup, should we check that the effective values for the HTTP2 timeouts and this grace period are coherent and warn if they are not?
@dchen1107 @liggitt Would it make sense to have this in a separate PR, or just update this one?

Comment (devppratik) [reply pkg/controller/nodelifecycle/config/v1alpha1/defaults.go:46]: Since I updated the flag description, I believe the schema description in zz_generated.openapi.go should have been updated when I ran `make update`. Is this the correct way or did I do something wrong?

File: pkg/controller/nodelifecycle/config/v1alpha1/defaults.go
Comment (rtheis) [inline pkg/controller/nodelifecycle/config/v1alpha1/defaults.go:46]: Do we need to update https://github.com/kubernetes/kubernetes/blob/2171bcb78966539a0c66b5ae8d9a96f9277a7e82/pkg/controller/nodelifecycle/node_lifecycle_controller_test.go#L55 to go along with this change?

Comment (devppratik) [reply pkg/controller/nodelifecycle/config/v1alpha1/defaults.go:46]: Make sense to update as the default is changed. Updated the value.

File: pkg/controller/nodelifecycle/config/v1alpha1/defaults.go
Comment (devppratik) [reply pkg/controller/nodelifecycle/config/v1alpha1/defaults.go:46]: Hello @dchen1107 @liggitt 
I updated the PR including why the change is required correctly.

Comment (dchen1107) [reply pkg/controller/nodelifecycle/config/v1alpha1/defaults.go:46]: LGTM and thanks!

**Comment by dchen1107** [comment]
/lgtm

cc/ @pacoxu

**Comment by pacoxu** [comment]
/lgtm
ping @liggitt for approve

BTW
- Do we want to open an exception for this change in v1.31 as code freeze now?
- Should this be cherry-picked to previous releases?

**Comment by liggitt** [comment]
/approve

For API doc updates. Since it's not a regression and is tunable, my inclination would be to not backport, but I'll defer to node leads on the milestone 1.31 question.

**Comment by devppratik** [comment]
@liggitt @dchen1107 @aojea Can the hold be removed or are we awaiting decision from node team before this is merged?

**Comment by aojea** [comment]
/hold cancel

I added the hold for Jordan and Dawn to comment, but they already approved after that.

**Review by rtheis — APPROVED** [review]
Thanks!

**Comment by devppratik** [comment]
> Do we want to open an exception for this change in v1.31 as code freeze now?

> For API doc updates. Since it's not a regression and is tunable, my inclination would be to not backport, but I'll defer to node leads on the milestone 1.31 question.

@dchen1107 @pacoxu Requesting your thoughts for the question regarding 1.31

**Comment by pacoxu** [comment]
Leave to sig node approvers to decide.
- pinged in slack https://kubernetes.slack.com/archives/C0BP8PW9G/p1721970518554739

**Comment by pacoxu** [comment]
Since this is tunable, we may merge this early in v1.32 release cycle IIUC.

/milestone v1.32

**Pull request merged**
